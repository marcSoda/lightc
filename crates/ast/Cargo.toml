[package]
name = "ast"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
serde = { version = "1.0.136", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }
