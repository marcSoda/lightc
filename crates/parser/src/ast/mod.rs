use std::fmt::Display;

use serde::Serialize;

pub mod prototype;
pub use prototype::Prototype;
pub mod stmt;
pub use stmt::*;
pub mod expr;
pub use expr::*;

#[derive(Debug, PartialEq, Serialize)]
pub struct Ast<T> {
    nodes: Vec<T>,
}

impl<T> Ast<T> {
    pub fn new() -> Self {
        Ast { nodes: vec![] }
    }

    pub fn add(&mut self, node: T) {
        self.nodes.push(node)
    }

    pub fn nodes(&self) -> &[T] {
        &self.nodes
    }

    pub fn into_nodes(self) -> Vec<T> {
        self.nodes
    }
}

impl<T> Default for Ast<T> {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum Node {
    Stmt(Statement),
    Expr(Expression),
}

impl Node {
    pub fn new<F, T>(cons: F, inner: T) -> Self
    where
        F: core::ops::Fn(T) -> Self,
    {
        (cons)(inner)
    }

    pub fn as_stmt(&self) -> &Statement {
        match self {
            Node::Stmt(s) => s,
            Node::Expr(_) => unreachable!("expected Statement"),
        }
    }

    pub fn to_expr(self) -> Expression {
        match self {
            Node::Stmt(_) => unreachable!("expected Expression"),
            Node::Expr(e) => e,
        }
    }

    pub fn as_expr(&self) -> &Expression {
        match self {
            Node::Stmt(_) => unreachable!("expected Expression"),
            Node::Expr(e) => e,
        }
    }

    pub fn as_expr_mut(&mut self) -> &mut Expression {
        match self {
            Node::Stmt(_) => unreachable!("expected Expression"),
            Node::Expr(e) => e,
        }
    }
}

impl Display for Node {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Node::Stmt(stmt) => write!(f, "{}", stmt),
            Node::Expr(expr) => write!(f, "{}", expr),
        }
    }
}

// Immutable visitor interface

pub trait AstVisitor {
    type Result;

    fn visit_stmt(&mut self, s: Statement) -> Self::Result;
    fn visit_expr(&mut self, e: Expression) -> Self::Result;
}

pub trait Visitable {
    fn accept<V: AstVisitor>(self, v: &mut V) -> V::Result;
}

impl Visitable for Node {
    fn accept<V: AstVisitor>(self, v: &mut V) -> V::Result {
        match self {
            Node::Stmt(s) => v.visit_stmt(s),
            Node::Expr(e) => v.visit_expr(e),
        }
    }
}

impl Visitable for Statement {
    fn accept<V: AstVisitor>(self, v: &mut V) -> V::Result {
        v.visit_stmt(self)
    }
}

impl Visitable for Expression {
    fn accept<V: AstVisitor>(self, v: &mut V) -> V::Result {
        v.visit_expr(self)
    }
}
