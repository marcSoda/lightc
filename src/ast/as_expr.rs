use super::*;

// As immutable expresions

pub(crate) trait AsExpr<'a> {
    type Item;

    fn as_expr(&'a self) -> Result<Self::Item, String>;
}

impl<'a> AsExpr<'a> for Node {
    type Item = &'a Expression;

    fn as_expr(&'a self) -> Result<Self::Item, String> {
        match self {
            Node::Stmt(_) => Err("Expected expresion in node. Found statement.".to_string()),
            Node::Expr(e) => Ok(e),
        }
    }
}

impl<'a> AsExpr<'a> for Box<Node> {
    type Item = &'a Expression;

    fn as_expr(&'a self) -> Result<Self::Item, String> {
        (**self).as_expr()
    }
}

impl<'a> AsExpr<'a> for Option<Box<Node>> {
    type Item = Option<&'a Expression>;

    fn as_expr(&'a self) -> Result<Self::Item, String> {
        self.as_deref().map(Node::as_expr).transpose()
    }
}

impl<'a> AsExpr<'a> for Vec<Node> {
    type Item = Vec<&'a Expression>;

    fn as_expr(&'a self) -> Result<Self::Item, String> {
        self.iter().map(|n| n.as_expr()).collect()
    }
}

impl<'a> AsExpr<'a> for Option<Vec<Node>> {
    type Item = Option<Vec<&'a Expression>>;

    fn as_expr(&'a self) -> Result<Self::Item, String> {
        self.as_ref().map(|o| o.as_expr()).transpose()
    }
}

// As mutable expresions

pub(crate) trait AsExprMut<'a> {
    type Item;

    fn as_expr_mut(&'a mut self) -> Result<Self::Item, String>;
}

impl<'a> AsExprMut<'a> for Node {
    type Item = &'a mut Expression;

    fn as_expr_mut(&'a mut self) -> Result<Self::Item, String> {
        match self {
            Node::Stmt(_) => Err("Expected expresion in node. Found statement.".to_string()),
            Node::Expr(e) => Ok(e),
        }
    }
}

impl<'a> AsExprMut<'a> for Option<Box<Node>> {
    type Item = Option<&'a mut Expression>;

    fn as_expr_mut(&'a mut self) -> Result<Self::Item, String> {
        self.as_deref_mut().map(Node::as_expr_mut).transpose()
    }
}

impl<'a> AsExprMut<'a> for Vec<Node> {
    type Item = Vec<&'a mut Expression>;

    fn as_expr_mut(&'a mut self) -> Result<Self::Item, String> {
        self.iter_mut().map(|n| n.as_expr_mut()).collect()
    }
}
