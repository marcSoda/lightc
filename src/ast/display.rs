use std::fmt::Display;

use super::{AstNode, Expression, Prototype, Function};

impl Display for AstNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AstNode::Expr(expr) => write!(f, "{}", expr),
            AstNode::Func(func) => write!(f, "{}", func),
        }
    }
}

impl Display for Expression {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use Expression::*;
        match self {
            U64(v) => write!(f, "{}", v),
            I64(v) => write!(f, "{}", v),
            F64(v) => write!(f, "{}", v),
            BinOp { sym, lhs, rhs } => write!(f, "({} {} {})", sym, lhs, rhs),
            UnOp { sym, rhs } => write!(f, "({} {})", sym, rhs),
            Var { name } => write!(f, "{}", name),
            Call { name, args } => {
                let mut s = format!("({}", name);
                if !args.is_empty() {
                    for arg in args {
                        s += &format!(" {}", arg);
                    }
                }
                write!(f, "{})", s)
            }
            Cond { cond, cons, alt } => {
                let mut s = format!("(if {}", cond);
                s += &cons.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });

                if let Some(alt) = alt {
                    s += &alt.iter().fold(String::new(), |mut acc, n| {
                        acc += &format!(" {}", n);
                        acc
                    });
                }
                write!(f, "{})", s)
            }
            For {
                var_name,
                start,
                cond,
                step,
                body,
            } => {
                let mut s = format!("(for (let {} {}) {} {}", var_name, start, cond, step);
                s += &body.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });
                write!(f, "{})", s)
            }
            Let { name, ty, init } => {
                let mut s = format!("(let {}:{}", name, ty);
                if let Some(body) = init {
                    s += &format!(" {}", body);
                }
                write!(f, "{})", s)
            }
        }
    }
}

impl Display for Prototype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut s = format!("({}", self.name);
        if !self.args.is_empty() {
            for arg in &self.args {
                s += &format!(" {}:{}", arg.0, arg.1);
            }
        }
        write!(f, "{})", s)
    }
}

impl Display for Function {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.body {
            Some(body) if !body.is_empty() => {
                let s = body.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });
                write!(f, "(define {}{})", self.proto, s)
            }
            _ => write!(f, "(define {})", self.proto),
        }
    }
}
