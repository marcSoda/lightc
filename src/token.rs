use serde::Serialize;

#[derive(PartialEq, Serialize)]
pub(crate) struct Token {
    pub(crate) tt: TokenType,
    pub(crate) line: usize,
    pub(crate) column: usize,
}

impl Token {
    // I believe the need for pub here, and also the warning, is a rust-analyzer
    // bug
    #[allow(dead_code)]
    pub(crate) fn new(tt: TokenType, line: usize, column: usize) -> Self {
        Token { tt, line, column }
    }
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.tt)
    }
}

impl std::fmt::Debug for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.tt)
    }
}

impl Default for Token {
    fn default() -> Self {
        Token {
            tt: TokenType::Eof,
            line: 0,
            column: 0,
        }
    }
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub(crate) enum TokenType {
    CloseBrace,
    CloseParen,
    Colon,
    Comma,
    Else,
    Eof,
    Extern,
    Fn,
    For,
    Ident(String),
    If,
    Let,
    Num(String),
    Op(Symbol),
    OpenBrace,
    OpenParen,
    Semicolon,
    VarType(Type),
}

impl std::fmt::Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenType::*;

        match self {
            Eof => write!(f, "EOF"),
            Op(s) => write!(f, "{}", s),
            Ident(i) => write!(f, "identifier: {}", i),
            Num(i) => write!(f, "number: {}", i),
            tt => write!(f, "{:?}", tt),
        }
    }
}

// A Symbol is an extra layer of abstraction between TokenType::Op() and the
// actual character. Convenient in Rust to help constrain matching.
#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub(crate) enum Symbol {
    Add,
    And,
    Assign,
    Div,
    Eq,
    Gt,
    Lt,
    Mul,
    Not,
    Or,
    Pow,
    RetType,
    Sub,
}

impl std::fmt::Display for Symbol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use Symbol::*;
        let s = match self {
            Add => "+",
            Assign => "=",
            And => "&&",
            Div => "/",
            Eq => "==",
            Gt => ">",
            Lt => "<",
            Mul => "*",
            Not => "!",
            Or => "||",
            Pow => "^",
            Sub => "-",
            RetType => "->",
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub(crate) enum Type {
    Int32,
    Int64,
    UInt32,
    UInt64,
    Float,
    Double,
    Void,
}

impl Default for Type {
    fn default() -> Self {
        Type::Void
    }
}

impl std::fmt::Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = format!("{:?}", self).to_ascii_lowercase();
        write!(f, "{}", s)
    }
}
