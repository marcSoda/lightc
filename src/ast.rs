use serde::Serialize;
use std::fmt::Display;

use crate::token::{Symbol, Type};

#[derive(Debug, PartialEq, Serialize)]
pub(crate) enum AstNode {
    Expr(Expression),
    Func(Function),
}

impl Display for AstNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AstNode::Expr(expr) => write!(f, "{}", expr),
            AstNode::Func(func) => write!(f, "{}", func),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) enum Expression {
    I64(u64),
    U64(u64),
    F64(f64),
    Var {
        name: String,
    },
    BinOp {
        sym: Symbol,
        lhs: Box<Expression>,
        rhs: Box<Expression>,
    },
    UnOp {
        sym: Symbol,
        rhs: Box<Expression>,
    },
    Call {
        name: String,
        args: Vec<Expression>,
    },
    Cond {
        cond: Box<Expression>,
        cons: Vec<Expression>,
        alt: Option<Vec<Expression>>,
    },
    For {
        var_name: String,
        start: Box<Expression>,
        cond: Box<Expression>,
        step: Box<Expression>,
        body: Vec<Expression>,
    },
    Let {
        name: String,
        ty: Type,
        init: Option<Box<Expression>>,
    },
}

impl Display for Expression {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use Expression::*;
        match self {
            U64(v) => write!(f, "{}", v),
            I64(v) => write!(f, "{}", v),
            F64(v) => write!(f, "{}", v),
            BinOp { sym, lhs, rhs } => write!(f, "({} {} {})", sym, lhs, rhs),
            UnOp { sym, rhs } => write!(f, "({} {})", sym, rhs),
            Var { name } => write!(f, "{}", name),
            Call { name, args } => {
                let mut s = format!("({}", name);
                if !args.is_empty() {
                    for arg in args {
                        s += &format!(" {}", arg);
                    }
                }
                write!(f, "{})", s)
            }
            Cond { cond, cons, alt } => {
                let mut s = format!("(if {}", cond);
                s += &cons.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });

                if let Some(alt) = alt {
                    s += &alt.iter().fold(String::new(), |mut acc, n| {
                        acc += &format!(" {}", n);
                        acc
                    });
                }
                write!(f, "{})", s)
            }
            For {
                var_name,
                start,
                cond,
                step,
                body,
            } => {
                let mut s = format!("(for (let {} {}) {} {}", var_name, start, cond, step);
                s += &body.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });
                write!(f, "{})", s)
            }
            Let { name, ty, init } => {
                let mut s = format!("(let {}:{}", name, ty);
                if let Some(body) = init {
                    s += &format!(" {}", body);
                }
                write!(f, "{})", s)
            }
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct Prototype {
    pub(crate) name: String,
    pub(crate) args: Vec<String>,
}

impl Display for Prototype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut s = format!("({}", self.name);
        if !self.args.is_empty() {
            for arg in &self.args {
                s += &format!(" {}", arg);
            }
        }
        write!(f, "{})", s)
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct Function {
    pub(crate) proto: Box<Prototype>,
    pub(crate) body: Option<Vec<Expression>>,
}

impl Display for Function {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.body {
            Some(body) if !body.is_empty() => {
                let s = body.iter().fold(String::new(), |mut acc, n| {
                    acc += &format!(" {}", n);
                    acc
                });
                write!(f, "(define {}{})", self.proto, s)
            }
            _ => write!(f, "(define {})", self.proto),
        }
    }
}
