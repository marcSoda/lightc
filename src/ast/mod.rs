use serde::Serialize;

use crate::token::{Symbol, Type};

mod display;

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct Ast<T: Visitable> {
    nodes: Vec<T>,
}

impl<T: Visitable> Ast<T> {
    pub(crate) fn new() -> Self {
        Ast { nodes: vec![] }
    }

    pub(crate) fn add(&mut self, node: T) {
        self.nodes.push(node)
    }

    pub(crate) fn nodes(&self) -> &Vec<T> {
        &self.nodes
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) enum Node {
    Stmt(Statement),
    Expr(Expression),
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) enum Statement {
    Cond {
        cond: Box<Node>,
        cons: Vec<Node>,
        alt: Option<Vec<Node>>,
    },
    For {
        var_name: String,
        start: Box<Node>,
        cond: Box<Node>,
        step: Box<Node>,
        body: Vec<Node>,
    },
    Let {
        name: String,
        ty: Type,
        init: Option<Box<Node>>,
    },
    Fn {
        proto: Box<Prototype>,
        body: Option<Vec<Node>>,
    },
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) enum Expression {
    I64(u64),
    U64(u64),
    F64(f64),
    Ident {
        name: String,
    },
    BinOp {
        sym: Symbol,
        lhs: Box<Node>,
        rhs: Box<Node>,
    },
    UnOp {
        sym: Symbol,
        rhs: Box<Node>,
    },
    Call {
        name: String,
        args: Vec<Node>,
    },
}

impl Expression {
    pub(crate) fn is_type(&self, ty: Type) -> bool {
        matches!(
            (self, ty),
            (Expression::U64(_), Type::U64)
                | (Expression::I64(_), Type::I64)
                | (Expression::F64(_), Type::F64)
        )
    }

    pub(crate) fn as_type(&self) -> Type {
        match self {
            Expression::I64(_) => Type::I64,
            Expression::U64(_) => Type::U64,
            Expression::F64(_) => Type::F64,
            _ => todo!(),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct Prototype {
    pub(crate) name: String,
    pub(crate) args: Vec<(String, Type)>,
    pub(crate) ret_type: Option<Type>,
}

pub(crate) trait AstVisitor {
    type Result;

    fn visit_stmt(&mut self, s: &Statement) -> Self::Result;
    fn visit_expr(&mut self, e: &Expression) -> Self::Result;
}

pub(crate) trait Visitable {
    fn accept<V: AstVisitor>(&self, v: &mut V) -> V::Result;
}

impl Visitable for Node {
    fn accept<V: AstVisitor>(&self, v: &mut V) -> V::Result {
        match self {
            Node::Stmt(s) => v.visit_stmt(s),
            Node::Expr(e) => v.visit_expr(e),
        }
    }
}

impl Visitable for Expression {
    fn accept<V: AstVisitor>(&self, v: &mut V) -> V::Result {
        v.visit_expr(self)
    }
}

impl Visitable for Statement {
    fn accept<V: AstVisitor>(&self, v: &mut V) -> V::Result {
        v.visit_stmt(self)
    }
}
