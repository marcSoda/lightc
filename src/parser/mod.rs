use std::{iter::Peekable, slice::Iter};

use self::{errors::ParseError, precedence::OpPrec};
use crate::ast::{Ast, Expression, Node, Prototype, Statement};
use crate::token::{Symbol, Token, TokenType, Type};

#[macro_use]
mod macros;
mod errors;
mod precedence;

type ParseResult = Result<Node, ParseError>;

pub(crate) struct Parser<'a> {
    ast: Ast<Node>,
    tokens: Peekable<Iter<'a, Token>>,
}

impl<'a> Parser<'a> {
    pub(crate) fn new(tokens: &'a [Token]) -> Self {
        Parser {
            ast: Ast::new(),
            tokens: tokens.iter().peekable(),
        }
    }

    // Parse each token using recursive descent
    pub(crate) fn parse(mut self) -> Result<Ast<Node>, ParseError> {
        while let Some(t) = self.tokens.peek() {
            let node = match t.tt {
                //_ => AstNode::Expr(self.parse_expression(0)?),
                _ => self.parse_statement()?,
            };
            self.ast.add(node);
        }
        Ok(self.ast)
    }

    fn parse_statement(&mut self) -> ParseResult {
        let token = self
            .tokens
            .peek()
            .ok_or_else(|| "Premature end of statement".to_string())?;

        let expr = match &token.tt {
            TokenType::If => self.parse_cond()?,
            TokenType::For => self.parse_for()?,
            TokenType::Let => self.parse_let()?,
            TokenType::Fn => self.parse_function()?,
            TokenType::Extern => self.parse_extern()?,
            _ => self.parse_expression(0)?,
        };
        Ok(expr)
    }

    // Consume token and dispatch. These are all considered primary expressions.
    //
    // TODO: for, let, if will likely become statements and need to be removed.
    fn parse_expression(&mut self, min_p: u8) -> ParseResult {
        let token = self
            .tokens
            .next()
            .ok_or_else(|| "Premature end of expression".to_string())?;

        let expr = match &token.tt {
            TokenType::Num(num) => self.parse_num(num, token)?,
            TokenType::Ident(id) => self.parse_ident(id)?,
            TokenType::OpenParen => self.parse_paren()?,
            TokenType::Op(sym) => self.parse_unop(*sym)?,
            x => {
                return Err(ParseError::from((
                    format!("Expecting primary expression. Got {}", x),
                    token,
                )))
            }
        };

        // Check for binop and process or return
        self.parse_binop(expr, min_p)
    }

    // Parses arbitrary length binary expressions. Uses Pratt with op
    // precedence parsing.
    fn parse_binop(&mut self, mut lhs: Node, min_p: u8) -> ParseResult {
        // Peek at the next token, otherwise return current lhs
        while let Some(next) = self.tokens.peek() {
            // Should always be an operator after a primary
            let sym = match next.tt {
                TokenType::Op(s) => s,
                // Start a new expression if we see two primaries in a row
                _ => break,
            };

            // Determine operator precedence and associativity.
            // Stop eating and return the lhs if the current op:
            //   - is lower precedence than the last one (min_p), or:
            //   - is the same precedence and associates left
            let p = match OpPrec::bin_prec(sym)? {
                OpPrec::Left(p) => {
                    if p <= min_p {
                        break;
                    }
                    p
                }
                OpPrec::Right(p) => {
                    if p < min_p {
                        break;
                    }
                    p
                }
            };

            // Advance past op
            self.tokens.next();

            // Descend for rhs with the current precedence as min_p
            let rhs = self.parse_expression(p)?;
            // Make a new lhs and continue loop
            lhs = Node::Expr(Expression::BinOp {
                sym,
                lhs: Box::new(lhs),
                rhs: Box::new(rhs),
            });
        }
        Ok(lhs)
    }

    fn parse_unop(&mut self, sym: Symbol) -> ParseResult {
        let p = OpPrec::un_prec(sym)?;
        let rhs = self.parse_expression(p)?;
        Ok(Node::Expr(Expression::UnOp {
            sym,
            rhs: Box::new(rhs),
        }))
    }

    fn parse_function(&mut self) -> ParseResult {
        // Eat 'fn'
        self.tokens.next();

        Ok(Node::Stmt(Statement::Fn {
            proto: Box::new(self.parse_proto()?),
            body: Some(self.parse_block()?),
        }))
    }

    fn parse_extern(&mut self) -> ParseResult {
        // Eat 'extern'
        self.tokens.next();

        Ok(Node::Stmt(Statement::Fn {
            proto: Box::new(self.parse_proto()?),
            body: None,
        }))
    }

    fn parse_proto(&mut self) -> Result<Prototype, ParseError> {
        let fn_name = expect_next_token!(
            self.tokens,
            TokenType::Ident(_),
            "Expecting function name in prototype"
        );

        expect_next_token!(
            self.tokens,
            TokenType::OpenParen,
            "Expecting '(' in prototype"
        );

        let mut args: Vec<(String, Type)> = vec![];
        while let Some(&next) = self.tokens.peek() {
            // Matches immediate ')'
            if next.tt == TokenType::CloseParen {
                break;
            }

            let id = expect_next_token!(
                self.tokens,
                TokenType::Ident(_),
                "Expecting ')' or identifier in prototype"
            );

            expect_next_token!(self.tokens, TokenType::Colon, "Expecting ':' in prototype");

            let ty = expect_next_token!(
                self.tokens,
                TokenType::VarType(_),
                "Expecting vartype in prototype"
            );

            args.push((id.to_string(), *ty));

            match self.tokens.peek() {
                Some(Token {
                    tt: TokenType::CloseParen,
                    ..
                }) => break,
                Some(Token {
                    tt: TokenType::Comma,
                    ..
                }) => self.tokens.next(), // Eat comma
                Some(x) => {
                    return Err(ParseError::from((
                        format!("Expecting ',' or ')' in prototype. Got {}", x),
                        *x,
                    )))
                }
                None => {
                    return Err(ParseError::from(String::from(
                        "Expecting one of [',', ')', ':']. Got EOF",
                    )))
                }
            };
        }

        // Eat close paren
        self.tokens.next();

        Ok(Prototype {
            name: fn_name.to_string(),
            args,
        })
    }

    fn parse_num(&self, num: &str, token: &Token) -> ParseResult {
        // if let Ok(n) = num.parse::<i64>() {
        //     Ok(Expression::I64(n))
        if let Ok(n) = num.parse::<u64>() {
            Ok(Node::Expr(Expression::U64(n)))
        } else if let Ok(n) = num.parse::<f64>() {
            Ok(Node::Expr(Expression::F64(n)))
        } else {
            Err(ParseError::from((
                format!("Invalid number literal: {}", token),
                token,
            )))
        }
    }

    // pub(crate) fn bool(s: &str) -> Option<Value> {
    //     match s.parse::<bool>() {
    //         Ok(b) => Some(Value::Bool(b)),
    //         Err(_) => None,
    //     }
    // }        Ok(Expression::Num { value: n })

    // Variable or function call
    fn parse_ident(&mut self, id: &str) -> ParseResult {
        let node = Expression::Ident {
            name: id.to_owned(),
        };

        // If next is not a '(', the current token is just a simple var
        match self.tokens.peek() {
            Some(Token {
                tt: TokenType::OpenParen,
                ..
            }) => (),
            _ => return Ok(Node::Expr(node)),
        };

        // Eat open paren
        self.tokens.next();

        let mut args: Vec<Node> = vec![];
        while let Some(&next) = self.tokens.peek() {
            // Matches immediate ')'
            if next.tt == TokenType::CloseParen {
                break;
            }

            args.push(self.parse_expression(0)?);

            match self.tokens.peek() {
                Some(Token {
                    tt: TokenType::CloseParen,
                    ..
                }) => break,
                Some(Token {
                    tt: TokenType::Comma,
                    ..
                }) => self.tokens.next(), // Eat comma
                _ => {
                    return Err(ParseError::from((
                        format!("Expecting ',' or ')' in function call. Got {}", next),
                        next,
                    )))
                }
            };
        }

        // Eat close paren
        expect_next_token!(
            self.tokens,
            TokenType::CloseParen,
            "Expecting ')' in function call"
        );

        Ok(Node::Expr(Expression::Call {
            name: id.to_owned(),
            args,
        }))
    }

    fn parse_paren(&mut self) -> ParseResult {
        let lhs = self.parse_expression(0);
        self.tokens.next(); // Eat ')'
        lhs
    }

    fn parse_cond(&mut self) -> ParseResult {
        self.tokens.next(); // Eat if

        let cond = self.parse_expression(0)?;
        let cons = self.parse_block()?;

        let alt = token_is_and_then!(self.tokens.peek(), TokenType::Else, {
            self.tokens.next(); // Eat else

            // To support `else if`, peek to check for `{` or `if`
            let next = self.tokens.peek();
            if matches!(next, Some(t) if t.tt != TokenType::If && t.tt != TokenType::OpenBrace) {
                return Err(ParseError::from((
                    "Expecting 'if' or '{' after else".to_string(),
                    *next.unwrap(),
                )));
            }

            let alt = self.parse_block()?;

            Some(alt)
        });

        Ok(Node::Stmt(Statement::Cond {
            cond: Box::new(cond),
            cons,
            alt,
        }))
    }

    fn parse_for(&mut self) -> ParseResult {
        self.tokens.next(); // Eat for

        // TODO: call parse_let() here
        expect_next_token!(self.tokens, TokenType::Let, "Expecting 'let' after for");

        let var_name = expect_next_token!(
            self.tokens,
            TokenType::Ident(_),
            "Expecting identifier after let"
        );
        expect_next_token!(
            self.tokens,
            TokenType::Op(Symbol::Assign),
            "Expecting '=' after identifer"
        );

        let start = self.parse_expression(0)?;
        expect_next_token!(
            self.tokens,
            TokenType::Semicolon,
            "Expecting ';' after start"
        );

        let cond = self.parse_expression(0)?;
        expect_next_token!(
            self.tokens,
            TokenType::Semicolon,
            "Expecting ';' after condition"
        );

        let step = self.parse_expression(0)?;

        Ok(Node::Stmt(Statement::For {
            var_name: var_name.to_owned(),
            start: Box::new(start),
            cond: Box::new(cond),
            step: Box::new(step),
            body: self.parse_block()?,
        }))
    }

    fn parse_let(&mut self) -> ParseResult {
        self.tokens.next(); // Eat let

        let var_name = expect_next_token!(
            self.tokens,
            TokenType::Ident(_),
            "Expecting identifier after let"
        );

        expect_next_token!(
            self.tokens,
            TokenType::Colon,
            "Expecting colon in let statement"
        );

        let ty = expect_next_token!(
            self.tokens,
            TokenType::VarType(_),
            "Type annotation required in let statement"
        );

        let init = token_is_and_then!(self.tokens.peek(), TokenType::Op(Symbol::Assign), {
            self.tokens.next();
            Some(self.parse_expression(0)?)
        });

        Ok(Node::Stmt(Statement::Let {
            name: var_name.to_owned(),
            ty: *ty,
            init: init.map(Box::new),
        }))
    }

    // Helper to collect a bunch of expressions enclosed by braces into a Vec<T>
    fn parse_block(&mut self) -> Result<Vec<Node>, ParseError> {
        let mut block: Vec<Node> = vec![];

        expect_next_token!(
            self.tokens,
            TokenType::OpenBrace,
            "Expecting '{' to start block"
        );

        while let Some(t) = self.tokens.peek() {
            match t.tt {
                TokenType::CloseBrace => {
                    self.tokens.next();
                    return Ok(block);
                }
                _ => block.push(self.parse_statement()?),
            }
        }

        // TODO: Could add more context here with some refactor
        Err(ParseError::from(
            "Expecting '}' to terminate block".to_string(),
        ))
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_yaml_snapshot, glob, with_settings};
    use std::{
        fs::File,
        io::{BufRead, BufReader},
    };

    use crate::{
        ast::{Ast, Node},
        lexer::Lexer,
        parser::{ParseError, Parser},
    };

    fn ast_to_string(ast: Result<&Ast<Node>, &ParseError>) -> String {
        match ast {
            Ok(ast) => ast.nodes().iter().map(|x| x.to_string()).collect(),
            Err(err) => err.to_string(),
        }
    }

    #[test]
    fn test_parser() {
        with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            glob!("tests/inputs/*.input", |path| {
                let file = File::open(path).expect("Error reading input file");
                let reader = BufReader::new(file);

                // Each line of the input files is meant to be a separate test
                // case. Treat each as its own AST. Including `ast_string` in the
                // output makes it more readable.
                let asts = reader
                    .lines()
                    .map(|line| {
                        let line = line.expect("Error reading input line");
                        let tokens = Lexer::new(&line).collect::<Result<Vec<_>, _>>().unwrap();
                        let ast = Parser::new(&tokens).parse();
                        let ast_string = ast_to_string(ast.as_ref());
                        (ast, ast_string)
                    })
                    .collect::<Vec<_>>();
                assert_yaml_snapshot!(asts);
            });
        });
    }
}
