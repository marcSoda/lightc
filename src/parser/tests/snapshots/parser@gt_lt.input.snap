---
source: src/parser/mod.rs
assertion_line: 555
expression: asts
input_file: src/parser/tests/inputs/gt_lt.input
---
- - a + b > 40
  - Ok:
      nodes:
        - Expr:
            BinOp:
              sym: Gt
              lhs:
                Expr:
                  BinOp:
                    sym: Add
                    lhs:
                      Expr:
                        Ident:
                          name: a
                          ty: ~
                    rhs:
                      Expr:
                        Ident:
                          name: b
                          ty: ~
                    ty: ~
              rhs:
                Expr:
                  Lit:
                    value:
                      UInt64: 40
                    ty: ~
              ty: ~
  - (> (+ a b) 40)
- - a < b / 40
  - Ok:
      nodes:
        - Expr:
            BinOp:
              sym: Lt
              lhs:
                Expr:
                  Ident:
                    name: a
                    ty: ~
              rhs:
                Expr:
                  BinOp:
                    sym: Div
                    lhs:
                      Expr:
                        Ident:
                          name: b
                          ty: ~
                    rhs:
                      Expr:
                        Lit:
                          value:
                            UInt64: 40
                          ty: ~
                    ty: ~
              ty: ~
  - (< a (/ b 40))

